arches/app/views/workflow_history.py

```diff
@@ -0,0 +1,75 @@
+import json
+
+from django.db import transaction
+from django.db.models import F, JSONField, Value
+from django.db.models.expressions import CombinedExpression
+from django.db.utils import IntegrityError
+from django.utils.translation import gettext as _
+from django.views.generic import View
+
+from arches.app.utils.permission_backend import user_is_resource_reviewer
+from arches.app.utils.response import JSONErrorResponse, JSONResponse
+from arches.app.models import models
+
+
+class WorkflowHistoryView(View):
+
+    def get(self, request, workflowid):
+        if not user_is_resource_reviewer(request.user):
+            return JSONErrorResponse(_("Request Failed"), _("Permission Denied"), status=403)
+        try:
+            workflow_history = models.WorkflowHistory.objects.get(workflowid=workflowid, user=request.user)
+        except models.WorkflowHistory.DoesNotExist:
+            workflow_history = {}
+        return JSONResponse(workflow_history, status=200)
+
+    def post(self, request, workflowid):
+        if not user_is_resource_reviewer(request.user):
+            return JSONErrorResponse(_("Request Failed"), _("Permission Denied"), status=403)
+
+        data = json.loads(request.body)
+        stepdata = data.get("stepdata", {})
+        componentdata = data.get("componentdata", {})
+
+        # TODO(Django 5.0) rewrite as a simpler update_or_create()
+        # call using different `defaults` vs. `create_defaults`
+        with transaction.atomic():
+            workflowid = data["workflowid"]
+            try:
+                history, created = models.WorkflowHistory.objects.select_for_update().get_or_create(
+                    workflowid = workflowid,
+                    user = request.user,
+                    defaults = {
+                        "stepdata": stepdata,
+                        "componentdata": componentdata,
+                        "user": request.user,
+                        "completed": data.get("completed", False),
+                    },
+                )
+            except IntegrityError:
+                # Wrong user for given workflowid.
+                return JSONErrorResponse(_("Request Failed"), _("Permission Denied"), status=403)
```

Comment:
Would be nice to add a test case for this, but noticed the test suite doesn't run `add_test_users` yet, so we don't have more than 1 users who is a resource reviewer. (I added add_test_users() to the afs tests, so it wouldn't be a big lift.)

Reply:
The UI should prevent you from getting into this scenario, given that we currently limit you to retrieving your own workflows.